crates/sui-framework/sources/governance/sui_system.move

```diff
@@ -569,6 +581,13 @@ module sui::sui_system {
         dynamic_field::borrow_mut(&mut self.id, self.version)
     }
 
+    fun upgrade_system_state(cur_state: SuiSystemStateInner): SuiSystemStateInner {
+        // Whenever we upgrade the system state version, we will have to first
+        // ship a framework upgrade that introduces a new system state type, and make this
+        // function generate such type from the old state.
```

Comment:
Also note that there will need to be a gap of an epoch where we are still using the old system state, i.e. the ordering is:

- Epoch N: Validators upgrade binary to include framework which has new system state type and migration function.
- Epoch N + k: Validators run framework upgrade at the end of `ChangeEpoch` out of this epoch
- Epoch N + k + 1: VM and caches reset at the beginning of epoch -- network "sees" changed framework.
- Epoch N + k + 1: run `update_system_state` with the new type during `advance_epoch` (i.e. "beginning" of `ChangeEpoch` out of this epoch), by incrementing `new_system_state_version`.

Is this manageable, or will this cause problems?  I can imagine that if there is a pressing bug in system state, you would want to be able to have a faster turnaround on calling the migration function, in which case you might want to consider running it at the beginning of the epoch following the framework upgrade rather than the end?

Reply:
Is there an alternative, though?

Reply:
> in which case you might want to consider running it at the beginning of the epoch following the framework upgrade rather than the end?

This is what I had in mind as the alternative, is it not feasible?

Reply:
Maybe? Below are the several challenges we need to solve:
1. In order to make this work, in the current design we need to go through two consecutive framework upgrades. In the first upgrade, we introduce the new system state type as well as the upgrade function; and then in the second upgrade, we change all internal functions to use the new type. If we only have one framework upgrade, we would then need to keep 2 copies of internal implementations on both types, and in every entry function, first look at the version before deciding which one to call.
2. We currently don't have a mechanism to inject a system transaction at the beginning of an epoch. This might not be too hard, but I need to think about what it means if we upgrade the system state type during the epoch.

I wonder if we should instead have a mechanism to make an epoch short.

Reply:
> If we only have one framework upgrade, we would then need to keep 2 copies of internal implementations on both types, and in every entry function, first look at the version before deciding which one to call.

Could you elaborate on this point?  I think I'm missing a detail on why the two versions of the entry function are required -- what I meant to suggest was for the new type, migration function, and changes to entry functions to all land in one upgrade and then the first transaction in the next epoch would run the migration function, so no other transaction could try and access the system state object and find that it hadn't been migrated.

> This might not be too hard, but I need to think about what it means if we upgrade the system state type during the epoch.

Was a bit lost here as well -- the aim of the new transaction at the beginning of an epoch is specifically to handle the presence of a new (inner) system state type, what interactions did you have in mind?  Maybe an important detail here is that because of compatibility rules, the new framework will still contain the old inner system state type, just as it was in the old framework, so the migration function would be able to access both.

> I wonder if we should instead have a mechanism to make an epoch short.

This would help, but it'd be interesting to hear the trade-off between this and injecting a system transaction at the beginning of the epoch

Reply:
> what I meant to suggest was for the new type, migration function, and changes to entry functions to all land in one upgrade and then the first transaction in the next epoch would run the migration function, so no other transaction could try and access the system state object and find that it hadn't been migrated.

Ah I see your point now. Indeed this would work.

On the second point, the primary challenge I see is whether/how we could guarantee that this upgrade transaction is run as the very first transaction in the epoch. And if it's not the very first, then how to reason about the interactions with other transactions that may be touching this object.

Another idea worth exploring is whether we could separate the framework upgrade tx and advance epoch tx, manually update the VM module cache right after the framework upgrade, and then execute the advance epoch tx to migrate it.

Reply:
I don't see why we need to have a new system transction in the new epoch. How would this be materially different from just adding another step to the implementation of ChangeEpochTx in execution_engine? We can just do the migration after calling advance_epoch, right?

Reply:
@lxfind and I chatted about this offline and came to a similar conclusion (that a new txn is not necessary except to cross the epoch boundary, to leverage reconfiguration to restart the VM and cache), but that being said, trying to restart those things outside of reconfiguration will very much grate against the grain.

Reply:
The migration can only happen after the framework upgrade, if we want to do everything in one epoch.
Then the VM module cache becomes tricky

Reply:
> Then the VM module cache becomes tricky

ah right, that's the catch

Reply:
programmable transactions with a move vm restart special command? :)